crates/telio-pmtu/src/linux.rs

```diff
@@ -0,0 +1,378 @@
+use std::{ffi::c_int, io, net::IpAddr, os::fd::AsRawFd, time::Duration};
+
+use pnet_packet::{icmp, icmpv6, Packet};
+use telio_utils::telio_log_debug;
+use tokio::net::UdpSocket;
+
+pub struct PMTUSocket {
+    sock: UdpSocket,
+    host: IpAddr,
+}
+
+const TIMEOUT: Duration = Duration::from_secs(1);
+const RETRIES: u32 = 3;
+
+impl PMTUSocket {
+    pub fn new(host: IpAddr) -> io::Result<Self> {
+        let sock = crate::create_icmp_socket(&host)?;
+        set_pmtu_on(&sock, host.is_ipv6())?;
+        Ok(Self { sock, host })
+    }
+
+    pub async fn probe_pmtu(&self) -> io::Result<u32> {
+        telio_log_debug!("Connecting to: {}", self.host);
+        self.sock.connect((self.host, 0)).await?;
+
+        telio_log_debug!("Connected to: {}", self.host);
+
+        let mut pmtu = read_pmtu(&self.sock, self.host.is_ipv6())?;
```

Comment:
If I am not mistaken, in case `read_mtu()` returns `Err(...)` we will silently fail. Maybe it is worth considering some `warn` or `info` level log in case of `read_pmtu()` failure in order to make debugging easier.

Reply:
I don't think it can fail just like that. But you're right, we cannot guarantee that it won't fail so let's handle this cleanly :+1:

Reply:
well, we probably will not be able to do anything too much about that error, than just report that PMTU detection has failed. But it would be nice to see in the logs what was exact reason for it failing, and here the logs come in :).

Reply:
Not necessarily, we can use some kind of default value if it fails. The PMTU read here is to provide a reasonable starting point for the search

Reply:
I did just that